File: akka-tutorials/akka-tutorial-second/src/main/java/akka/tutorial/java/second/Pi.java
/**
 * Copyright (C) 2009-2011 Scalable Solutions AB <http://scalablesolutions.se>
 */

package akka.tutorial.java.second;

import static akka.actor.Actors.actorOf;
import static akka.actor.Actors.poisonPill;
import static java.lang.System.currentTimeMillis;
import static java.util.Arrays.asList;
import scala.Option;
import akka.actor.ActorRef;
import akka.actor.Channel;
import akka.actor.UntypedActor;
import akka.actor.UntypedActorFactory;
import akka.dispatch.Future;
import akka.japi.Procedure;
import akka.routing.CyclicIterator;
import akka.routing.InfiniteIterator;
import akka.routing.Routing.Broadcast;
import akka.routing.UntypedLoadBalancer;

public class Pi {

  public static void main(String[] args) throws Exception {
    Pi pi = new Pi();
    pi.calculate(4, 10000, 10000);
  }

  // ====================
  // ===== Messages =====
  // ====================
  static class Calculate {}

  static class Work {
    private final int arg;
    private final int nrOfElements;

    public Work(int arg, int nrOfElements) {
      this.arg = arg;
      this.nrOfElements = nrOfElements;
    }

    public int getArg() { return arg; }
    public int getNrOfElements() { return nrOfElements; }
  }

  static class Result {
    private final double value;

    public Result(double value) {
      this.value = value;
    }

    public double getValue() { return value; }
  }

  // ==================
  // ===== Worker =====
  // ==================
  static class Worker extends UntypedActor {

    // define the work
    private double calculatePiFor(int arg, int nrOfElements) {
      double acc = 0.0;
      for (int i = arg * nrOfElements; i <= ((arg + 1) * nrOfElements - 1); i++) {
        acc += 4 * Math.pow(-1, i) / (2 * i + 1);
      }
      return acc;
    }

    // message handler
    public void onReceive(Object message) {
      if (message instanceof Work) {
        Work work = (Work) message;
        getContext().replyUnsafe(new Result(calculatePiFor(work.getArg(), work.getNrOfElements()))); // perform the work
      } else throw new IllegalArgumentException("Unknown message [" + message + "]");
    }
  }

  // ==================
  // ===== Master =====
  // ==================
  static class Master extends UntypedActor {
    private final int nrOfMessages;
    private final int nrOfElements;

    private double pi;
    private int nrOfResults;

    private ActorRef router;

    static class PiRouter extends UntypedLoadBalancer {
      private final InfiniteIterator<ActorRef> workers;

      public PiRouter(ActorRef[] workers) {
        this.workers = new CyclicIterator<ActorRef>(asList(workers));
      }

      public InfiniteIterator<ActorRef> seq() {
        return workers;
      }
    }

    public Master(int nrOfWorkers, int nrOfMessages, int nrOfElements) {
      this.nrOfMessages = nrOfMessages;
      this.nrOfElements = nrOfElements;

      // create the workers
      final ActorRef[] workers = new ActorRef[nrOfWorkers];
      for (int i = 0; i < nrOfWorkers; i++) {
        workers[i] = actorOf(Worker.class).start();
      }

      // wrap them with a load-balancing router
      router = actorOf(new UntypedActorFactory() {
        public UntypedActor create() {
          return new PiRouter(workers);
        }
      }).start();
    }

    @Override
    public void preStart() {
      become(scatter);
    }

    // message handler
    public void onReceive(Object message) {
      throw new IllegalStateException("Should be gather or scatter");
    }

    private final Procedure<Object> scatter = new Procedure<Object>() {
      public void apply(Object msg) {
        // schedule work
        for (int arg = 0; arg < nrOfMessages; arg++) {
          router.tell(new Work(arg, nrOfElements), getContext());
        }
        // Assume the gathering behavior
        become(gather(getContext().getChannel()));
      }
    };

    private Procedure<Object> gather(final Channel<Object> recipient) {
      return new Procedure<Object>() {
        public void apply(Object msg) {
          // handle result from the worker
          Result result = (Result) msg;
          pi += result.getValue();
          nrOfResults += 1;
          if (nrOfResults == nrOfMessages) {
            // send the pi result back to the guy who started the calculation
            recipient.tell(pi);
            // shut ourselves down, we're done
            getContext().stop();
          }
        }
      };
    }

    @Override
    public void postStop() {
      // send a PoisonPill to all workers telling them to shut down themselves
      router.tell(new Broadcast(poisonPill()));
      // send a PoisonPill to the router, telling him to shut himself down
      router.tell(poisonPill());
    }
  }

  // ==================
  // ===== Run it =====
  // ==================
  public void calculate(final int nrOfWorkers, final int nrOfElements, final int nrOfMessages) throws Exception {

    // create the master
    ActorRef master = actorOf(new UntypedActorFactory() {
      public UntypedActor create() {
        return new Master(nrOfWorkers, nrOfMessages, nrOfElements);
      }
    }).start();

    // start the calculation
    long start = currentTimeMillis();

    // send calculate message
    long timeout = 60000;
    Future<Object> replyFuture = master.sendRequestReplyFuture(new Calculate(), timeout, null);
    Option<Object> result = replyFuture.await().resultOrException();
    if (result.isDefined()) {
      double pi = (Double) result.get();
      // TODO java api for EventHandler?
//      EventHandler.info(this, String.format("\n\tPi estimate: \t\t%s\n\tCalculation time: \t%s millis", pi, (currentTimeMillis() - start)));
      System.out.println(String.format("\n\tPi estimate: \t\t%s\n\tCalculation time: \t%s millis", pi, (currentTimeMillis() - start)));
    } else {
      // TODO java api for EventHandler?
//      EventHandler.error(this, "Pi calculation did not complete within the timeout.");
      System.out.println("Pi calculation did not complete within the timeout.");
    }

  }
}

File: akka-tutorials/akka-tutorial-second/src/main/scala/Pi.scala
/**
 * Copyright (C) 2009-2011 Typesafe Inc. <http://www.typesafe.com>
 */

package akka.tutorial.second

import akka.actor.Actor._
import akka.event.EventHandler
import System.{ currentTimeMillis ⇒ now }
import akka.routing.Routing.Broadcast
import akka.routing._
import akka.AkkaApplication
import akka.actor.{ ActorRef, Timeout, Actor, PoisonPill }

object Pi extends App {

  val app = AkkaApplication()

  calculate(nrOfWorkers = 4, nrOfElements = 10000, nrOfMessages = 10000)

  // ====================
  // ===== Messages =====
  // ====================
  sealed trait PiMessage
  case object Calculate extends PiMessage
  case class Work(arg: Int, nrOfElements: Int) extends PiMessage
  case class Result(value: Double) extends PiMessage

  // ==================
  // ===== Worker =====
  // ==================
  class Worker() extends Actor {
    // define the work
    val calculatePiFor = (arg: Int, nrOfElements: Int) ⇒ {
      val range = (arg * nrOfElements) to ((arg + 1) * nrOfElements - 1)
      var acc = 0.0D
      range foreach (i ⇒ acc += 4 * math.pow(-1, i) / (2 * i + 1))
      acc
      //range map (j => 4 * math.pow(-1, j) / (2 * j + 1)) sum
    }

    def receive = {
      case Work(arg, nrOfElements) ⇒ sender ! Result(calculatePiFor(arg, nrOfElements)) // perform the work
    }
  }

  // ==================
  // ===== Master =====
  // ==================
  case class Master(nrOfWorkers: Int, nrOfElements: Int, nrOfMessages: Int) extends Actor {
    var pi: Double = _
    var nrOfResults: Int = _

    // create the workers
    val workers = Vector.fill(nrOfWorkers)(app.actorOf[Worker])

    // wrap them with a load-balancing router
    val router = app.actorOf(RoutedProps(
      routerFactory = () ⇒ new RoundRobinRouter,
      connectionManager = new LocalConnectionManager(workers)), "pi")

    // phase 1, can accept a Calculate message
    def scatter: Receive = {
      case Calculate ⇒
        // schedule work
        for (arg ← 0 until nrOfMessages) router ! Work(arg, nrOfElements)

        //Assume the gathering behavior
        this become gather(sender)
    }

    // phase 2, aggregate the results of the Calculation
    def gather(recipient: ActorRef): Receive = {
      case Result(value) ⇒
        // handle result from the worker
        pi += value
        nrOfResults += 1
        if (nrOfResults == nrOfMessages) {
          // send the pi result back to the guy who started the calculation
          recipient ! pi
          // shut ourselves down, we're done
          self.stop()
        }
    }

    // message handler starts at the scattering behavior
    def receive = scatter

    // when we are stopped, stop our team of workers and our router
    override def postStop() {
      // send a PoisonPill to all workers telling them to shut down themselves
      router ! Broadcast(PoisonPill)
      // send a PoisonPill to the router, telling him to shut himself down
      router ! PoisonPill
    }
  }

  // ==================
  // ===== Run it =====
  // ==================
  def calculate(nrOfWorkers: Int, nrOfElements: Int, nrOfMessages: Int) {
    // create the master
    val master = app.actorOf(new Master(nrOfWorkers, nrOfElements, nrOfMessages))

    //start the calculation
    val start = now

    //send calculate message
    master.?(Calculate, Timeout(60000)).
      await.resultOrException match { //wait for the result, with a 60 seconds timeout
        case Some(pi) ⇒
          app.eventHandler.info(this, "\n\tPi estimate: \t\t%s\n\tCalculation time: \t%s millis".format(pi, (now - start)))
        case None ⇒
          app.eventHandler.error(this, "Pi calculation did not complete within the timeout.")
      }
  }
}
